src/GafferUI/NoduleLayout.cpp

```diff
@@ -0,0 +1,590 @@
+	{
+		return b->readable();
+	}
+
+	return true;
+}
+
+} // namespace
+
+//////////////////////////////////////////////////////////////////////////
+// NoduleLayout implementation
+//////////////////////////////////////////////////////////////////////////
+
+IE_CORE_DEFINERUNTIMETYPED( NoduleLayout );
+
+NoduleLayout::NoduleLayout( Gaffer::GraphComponentPtr parent, IECore::InternedString section )
+	:	Gadget(), m_parent( parent ), m_section( section )
+{
+	LinearContainerPtr noduleContainer = new LinearContainer(
+		"__noduleContainer",
+		orientation( m_parent.get(), m_section ),
+		LinearContainer::Centre,
+		spacing( m_parent.get(), m_section ),
+		direction( m_parent.get(), m_section )
+	);
+
+	addChild( noduleContainer );
+
+	m_parent->childAddedSignal().connect( boost::bind( &NoduleLayout::childAdded, this, ::_2 ) );
+	m_parent->childRemovedSignal().connect( boost::bind( &NoduleLayout::childRemoved, this, ::_2 ) );
+
+	Metadata::plugValueChangedSignal().connect( boost::bind( &NoduleLayout::plugMetadataChanged, this, ::_1, ::_2, ::_3, ::_4 ) );
+	Metadata::nodeValueChangedSignal().connect( boost::bind( &NoduleLayout::nodeMetadataChanged, this, ::_1, ::_2, ::_3 ) );
+
+	updateNoduleLayout();
+}
+
+NoduleLayout::~NoduleLayout()
+{
+}
+
+Nodule *NoduleLayout::nodule( const Gaffer::Plug *plug )
+{
+	const GraphComponent *plugParent = plug->parent<GraphComponent>();
+	if( !plugParent )
+	{
+		return NULL;
+	}
+
+	if( plugParent == m_parent.get() )
+	{
+		NoduleMap::iterator it = m_nodules.find( plug );
+		if( it != m_nodules.end() )
+		{
+			return it->second.nodule.get();
+		}
+		return NULL;
+	}
+	else if( const Plug *parentPlug = IECore::runTimeCast<const Plug>( plugParent ) )
+	{
+		CompoundNodule *compoundNodule = IECore::runTimeCast<CompoundNodule>( nodule( parentPlug ) );
+		if( compoundNodule )
+		{
+			return compoundNodule->nodule( plug );
+		}
+	}
+	return NULL;
+}
+
+const Nodule *NoduleLayout::nodule( const Gaffer::Plug *plug ) const
+{
+	// naughty cast is better than repeating the above logic.
+	return const_cast<NoduleLayout *>( this )->nodule( plug );
+}
+
+LinearContainer *NoduleLayout::noduleContainer()
+{
+	return getChild<LinearContainer>( 0 );
+}
+
+const LinearContainer *NoduleLayout::noduleContainer() const
+{
+	return getChild<LinearContainer>( 0 );
+}
+
+void NoduleLayout::childAdded( Gaffer::GraphComponent *child )
+{
+	if( IECore::runTimeCast<Gaffer::Plug>( child ) )
+	{
+		updateNoduleLayout();
+	}
+}
+
+void NoduleLayout::childRemoved( Gaffer::GraphComponent *child )
+{
+	if( IECore::runTimeCast<Gaffer::Plug>( child ) )
+	{
+		updateNoduleLayout();
+	}
+}
+
+void NoduleLayout::plugMetadataChanged( IECore::TypeId nodeTypeId, const Gaffer::MatchPattern &plugPath, IECore::InternedString key, const Gaffer::Plug *plug )
+{
+	if( childAffectedByChange( m_parent.get(), nodeTypeId, plugPath, plug ) )
+	{
+		if(
+			key == g_sectionKey || key == g_indexKey || key == g_visibleKey ||
+			key == g_noduleTypeKey ||
```

Comment:
Sorry, I composed a reply in my head when I got the email, but this one didn't seem to make it out onto the page. The first line is layout metadata, the second line is nodule metadata, and the last line is deprecated stuff that'll be deleted one day. I don't think it's worth worrying about.